src/ripple/overlay/impl/Handshake.cpp

```diff
@@ -0,0 +1,302 @@
+    // is 0. Don't allow this.
+    if (result == beast::zero)
+    {
+        JLOG(journal.error()) << "Cookie generation: identical finished messages";
+        return boost::none;
+    }
+
+    return sha512Half (Slice (result.data(), result.size()));
+}
+
+void
+buildHandshake(
+    boost::beast::http::fields& h,
+    const ripple::uint256& sharedValue,
+    boost::optional<std::uint32_t> networkID,
+    beast::IP::Address public_ip,
+    beast::IP::Address remote_ip,
+    Application& app)
+{
+    if (networkID)
+    {
+        // The network identifier, if configured, can be used to specify
+        // what network we intend to connect to and detect if the remote
+        // end connects to the same network.
+        h.insert("Network-ID", std::to_string(*networkID));
+    }
+
+    h.insert ("Network-Time",
+        std::to_string(app.timeKeeper().now().time_since_epoch().count()));
+
+    h.insert ("Public-Key",
+        toBase58(TokenType::NodePublic, app.nodeIdentity().first));
+
+    {
+        auto const sig = signDigest(app.nodeIdentity().first,
+            app.nodeIdentity().second, sharedValue);
+        h.insert("Session-Signature",
+            base64_encode(sig.data(), sig.size()));
+    }
+
+    if (beast::IP::is_public (remote_ip))
+        h.insert ("Remote-IP", remote_ip.to_string());
+
+    if (!public_ip.is_unspecified())
+        h.insert ("Local-IP", public_ip.to_string());
+
+    if (auto const cl = app.getLedgerMaster().getClosedLedger())
+    {
+        // TODO: Use hex for these
+        h.insert ("Closed-Ledger", base64_encode(
+            cl->info().hash.begin(), cl->info().hash.size()));
+        h.insert ("Previous-Ledger", base64_encode(
+            cl->info().parentHash.begin(), cl->info().parentHash.size()));
+    }
+}
+
+PublicKey
+verifyHandshake(
+    boost::beast::http::fields const& headers,
+    const ripple::uint256& sharedValue,
+    boost::optional<std::uint32_t> networkID,
+    beast::IP::Address public_ip,
+    beast::IP::Address remote,
+    Application& app)
+{
+    if (networkID)
+    {
+        auto const iter = headers.find("Network-ID");
+
+        if (iter != headers.end())
+        {
+            std::uint32_t nid;
+
+            if (!beast::lexicalCastChecked(nid, iter->value().to_string()))
+                throw std::runtime_error("Invalid peer network identifier");
+
+            if(nid != *networkID)
+                throw std::runtime_error("Peer is on a different network");
+        }
+    }
+
+    {
+        auto const iter = headers.find("Network-Time");
+
+        if (iter != headers.end())
+        {
+            auto const netTime = [str=iter->value().to_string()]()
+            {
+                boost::optional<TimeKeeper::time_point> ret;
+
+                TimeKeeper::duration::rep val;
+
+                if (beast::lexicalCastChecked(val, str))
+                    ret.emplace(TimeKeeper::duration{val});
+
+                return ret;
+            }();
```

Comment:
Could move the `throw` inside the lambda and get rid of the `optional`. I'd also put a return type on the lambda, since with the re-write it's not immediately obviously what the return type is Something like:
```
            auto const netTime =
                    [&iter]() -> TimeKeeper::time_point {
                TimeKeeper::duration::rep val;
                if (beast::lexicalCastChecked(val, iter->value().to_string()))
                    return TimeKeeper::time_point{TimeKeeper::duration{val}};
                // It's not an error for the header field to not be present but if
                // it is present and it contains junk data, that is an error.
                throw std::runtime_error("Invalid peer clock timestamp");
            }();
```